engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Array.java

```diff
@@ -179,21 +180,44 @@ boolean hasWarnings(@Shared("warnsLib") @CachedLibrary(limit = "3") WarningsLibr
 
   @ExportMessage
   Warning[] getWarnings(
-      Node location, @Shared("warnsLib") @CachedLibrary(limit = "3") WarningsLibrary warnings)
+      Node location,
+      boolean shouldWrap,
+      @Shared("warnsLib") @CachedLibrary(limit = "3") WarningsLibrary warnings)
       throws UnsupportedMessageException {
-    if (cachedWarnings == null) {
-      cachedWarnings = Warning.fromSetToArray(collectAllWarnings(warnings, location));
+    Warning[] cache = shouldWrap ? cachedWarningsWrapped : cachedWarningsUnwrapped;
+    if (cache == null) {
+      cache = Warning.fromSetToArray(collectAllWarnings(warnings, location, shouldWrap));
+      if (shouldWrap) {
+        cachedWarningsWrapped = cache;
+      } else {
+        cachedWarningsUnwrapped = cache;
+      }
     }
-    return cachedWarnings;
+    return cache;
   }
 
   @CompilerDirectives.TruffleBoundary
-  private EconomicSet<Warning> collectAllWarnings(WarningsLibrary warnings, Node location)
+  private EconomicSet<Warning> collectAllWarnings(
+      WarningsLibrary warningsLib, Node location, boolean shouldWrap)
       throws UnsupportedMessageException {
     EconomicSet<Warning> setOfWarnings = EconomicSet.create(new WithWarnings.WarningEquivalence());
-    for (Object item : items) {
-      if (warnings.hasWarnings(item)) {
-        setOfWarnings.addAll(Arrays.asList(warnings.getWarnings(item, location)));
+    for (int i = 0; i < this.items.length; i++) {
+      final int finalIndex = i;
+      Object item = this.items[i];
+      if (warningsLib.hasWarnings(item)) {
+        Warning[] warnings = warningsLib.getWarnings(item, location, shouldWrap);
+        Warning[] wrappedWarningsMaybe;
+
+        if (shouldWrap) {
+          wrappedWarningsMaybe =
+              Arrays.stream(warnings)
+                  .map(warning -> Warning.wrapMapError(warningsLib, warning, finalIndex))
+                  .toArray(Warning[]::new);
+        } else {
+          wrappedWarningsMaybe = warnings;
+        }
+
+        setOfWarnings.addAll(Arrays.asList(wrappedWarningsMaybe));
```

Comment:
Why is this called `...Maybe`? I don't see ever holding a `null`/None.

Reply:
Because if shouldWrap==False, it's not actually wrapped.